File: domain/src/main/java/com/example/domain/usecase/auth/LoginUseCase.kt
package com.example.domain.usecase.auth

import android.util.Log
import com.example.core_common.result.CustomResult
import com.example.domain.model.data.UserSession
import com.example.domain.model.enum.UserAccountStatus
import com.example.domain.model.vo.DocumentId
import com.example.domain.repository.base.AuthRepository
import com.example.domain.repository.base.UserRepository
import com.example.domain.usecase.user.GetUserUseCase
import kotlinx.coroutines.flow.first
import javax.inject.Inject

/**
 * 이메일과 비밀번호로 로그인을 시도하고 사용자 세션 정보를 반환하는 UseCase.
 * (기존 LoginUseCase를 수정함)
 *
 * @property authRepository 인증 관련 기능을 제공하는 Repository.
 */
/**
 * Custom exception to indicate that a user account has been withdrawn.
 */
class WithdrawnAccountException(message: String) : Exception(message)

class LoginUseCase @Inject constructor(
    private val authRepository: AuthRepository,
    private val getUserUseCase: GetUserUseCase,
    private val userRepository: UserRepository
) {
    // Centralized TAG for consistent Logcat filtering during debugging
    companion object {
        private const val TAG = "LoginUseCase"
    }

    /**
     * 이메일과 비밀번호를 이용하여 사용자 로그인을 시도합니다.
     *
     * @param credentials EmailPasswordCredentials 객체 (email, password 포함).
     * @return 성공 시 UserSession이 포함된 CustomResult.Success, 실패 시 CustomResult.Error.
     */
    suspend operator fun invoke(email : String, password : String): CustomResult<UserSession, Exception> {
        // 시작 로그 – 입력된 이메일 기준으로 로그인 시도
        Log.d(TAG, "Attempting login for email=$email")
        return when (val loginResult = authRepository.login(email, password)) {
            is CustomResult.Success -> {
                val userSession = loginResult.data
                // AuthRepository 에서 세션 획득 성공
                Log.d(TAG, "Auth success. Session=$userSession")
                // After successful authentication, fetch user details to check account status
                Log.d(TAG, "Fetching user details (may hit cache) for userId=${userSession.userId}")
                when (val userDetailsResult = getUserUseCase(userSession.userId).first()) {
                    is CustomResult.Success -> {
                        val user = userDetailsResult.data
                        // Firestore 에서 가져온 사용자 정보 (캐시 또는 서버)
                        Log.d(TAG, "Fetched User (possibly from cache): $user")
                        Log.d(TAG, "Account Status: ${user}")
                        if (user.accountStatus == UserAccountStatus.WITHDRAWN) {
                            Log.d(TAG, "Cached status=WITHDRAWN, verifying against server...")
                            when (val userResult = userRepository.observe(DocumentId(userSession.userId)).first()) {
                                is CustomResult.Success -> {
                                    val remoteUser = userResult.data
                                    Log.d(TAG, "Server status=${remoteUser.accountStatus}")
                                    if (remoteUser.accountStatus == UserAccountStatus.WITHDRAWN) {
                                        authRepository.logout()
                                        CustomResult.Failure(WithdrawnAccountException("탈퇴한 계정입니다. (서버 확인)"))
                                    } else {
                                        // Remote says ACTIVE -> stale cache, continue login success
                                        loginResult
                                    }
                                }
                                is CustomResult.Failure -> {
                                    authRepository.logout()
                                    CustomResult.Failure(userResult.error)
                                }
                                else -> {
                                    authRepository.logout()
                                    CustomResult.Failure(Exception("알 수 없는 오류 (원격 사용자 확인)"))
                                }
                            }
                        } else {
                            // Account is active, return the original success result with session
                            loginResult
                        }
                    }
                    is CustomResult.Failure -> {
                        // Failed to fetch user details – sign out to avoid stale session and propagate failure
                        authRepository.logout()
                        CustomResult.Failure(userDetailsResult.error ?: Exception("사용자 정보를 가져오는데 실패했습니다."))
                    }
                    else -> {
                        Log.d(TAG, "Unexpected userDetailsResult state: $userDetailsResult")
                        CustomResult.Failure(Exception("알 수 없는 오류가 발생했습니다."))
                    }
                }
            }
            is CustomResult.Failure -> {
                // Login authentication failed, log failure reason
                Log.d(TAG, "Authentication failed: ${loginResult.error}")
                // Login authentication failed, return the original failure
                loginResult
            }
            else -> {
                 CustomResult.Failure(Exception("알 수 없는 로그인 오류가 발생했습니다."))
            }
        }
    }
}

File: functions/src/domain/usecases/auth/session/LoginUseCase.ts
/**
 * Login use case
 * Handles user authentication with email and password
 */

import { Result } from '../../../../shared/types/Result';
import { FunctionError, FunctionErrorCode, UseCaseContext } from '../../../../shared/types/common';
import { AuthRepository, UserRepository } from '../../../repositories';
import { UserSession } from '../../../models/data/UserSession';
import { UserEmail, UserId } from '../../../models/vo';
import { UserAccountStatus } from '../../../models/enums/UserAccountStatus';
import { User } from '../../../models/base/User';

export class WithdrawnAccountError extends Error {
  constructor(message: string = 'Account has been withdrawn') {
    super(message);
    this.name = 'WithdrawnAccountError';
  }
}

export class LoginUseCase {
  constructor(
    private readonly authRepository: AuthRepository,
    private readonly userRepository: UserRepository
  ) {}

  async execute(
    email: UserEmail,
    password: string,
    context?: UseCaseContext
  ): Promise<Result<UserSession, Error>> {
    try {
      // Step 1: Attempt authentication
      const authResult = await this.authRepository.login(email, password);
      
      if (!authResult.isSuccess) {
        return authResult;
      }

      const userSession = authResult.getOrThrow();

      // Step 2: Verify user account status
      const userResult = await this.userRepository.findById(userSession.userId);
      
      if (!userResult.isSuccess) {
        // If user not found, logout and return error
        await this.authRepository.logout();
        return Result.failure(userResult.getOrNull() || new Error('User not found'));
      }

      const user = userResult.getOrThrow();

      // Step 3: Check account status
      if (user.accountStatus === UserAccountStatus.WITHDRAWN) {
        await this.authRepository.logout();
        return Result.failure(new WithdrawnAccountError('탈퇴한 계정입니다'));
      }

      if (user.accountStatus === UserAccountStatus.SUSPENDED) {
        await this.authRepository.logout();
        return Result.failure(new FunctionError(
          FunctionErrorCode.FORBIDDEN,
          '정지된 계정입니다',
          { userId: user.userId.value, status: user.accountStatus }
        ));
      }

      if (user.accountStatus === UserAccountStatus.LOCKED) {
        await this.authRepository.logout();
        return Result.failure(new FunctionError(
          FunctionErrorCode.FORBIDDEN,
          '잠긴 계정입니다. 관리자에게 문의하세요',
          { userId: user.userId.value, status: user.accountStatus }
        ));
      }

      if (!user.canLogin()) {
        await this.authRepository.logout();
        return Result.failure(new FunctionError(
          FunctionErrorCode.FORBIDDEN,
          '로그인할 수 없는 계정 상태입니다',
          { userId: user.userId.value, status: user.accountStatus, emailVerified: user.emailVerified }
        ));
      }

      // Step 4: Update last login time
      await this.userRepository.updateLastLogin(user.userId, new Date());

      // Step 5: Return successful session
      return Result.success(userSession);

    } catch (error) {
      return Result.failure(error instanceof Error ? error : new Error(String(error)));
    }
  }
}